### PR TITLE
Issue-941 dockerize-weird-binaries

### PR DESCRIPTION
**Pull request by flooie**
*** This is a draft *** because it will require a change to the docker-compose file before merging.  I also was hoping to get feedback to help fine-tune it.  

This is a moderately sized PR - to remove and isolate unwieldy code from CL itself.  

New docker containers added for Judge Pics (JP), Seal Rookery(SR) and Binary Transformer and Extractors (BTE).

JR and SR are just large and unwieldy installs that require a lot of time and effort to build everytime we download or build cl-django.  

BTE - is a new micro service that we call to transform audio, OCR and extract content from documents and generate thumbnails.  This removes all subprocess calls from the Django app.

Additionally, the celery docker image now inherits from Django to reduce redundant installs.

### CODE REVIEW

File: cl/scrapers/tasks.py [hunk 1]
```diff
@@ -567,51 +273,36 @@ def process_audio_file(pk):
...
-            "process_audio_file for item: %s\n"
-            "Traceback:\n"
-            "%s" % (af.pk, traceback.format_exc())
-        )
-        print(msg)
+    err = response["err"]
+    if not err == Timeout:
+        with open(tmp_path, "w") as mp3:
+            mp3.write(response["content"])
+        set_mp3_meta_data(af, tmp_path)
```
Comment (mlissner) [inline cl/scrapers/tasks.py:282]: Alas, everything before now was style, but I think this is feature change. I think all the eyed3 junk and metadata setting/tweaking junk should go into the transformer too. Wouldn't it be nice if we set a wav file and some metadata to the transformer and got back an MP3 with album art and good metadata that we can just save and get on with it?

Or if we want to be less monolithic about it, we could instead do two calls, one for "make an MP3" and one for "set its metdata". Either way though, I think eyed3 is a great candidate for yanking out of here.

File: cl/scrapers/tests.py [hunk 2]
```diff
@@ -438,3 +443,41 @@ def test_process_audio_file(self):
...
         )
+
+
+class BinaryTransformerExtractionTest(TestCase):
+    def setUp(self):
+        self.path = os.path.join(settings.MEDIA_ROOT, "test", "search")
+
+    def test_heartbeat(self):
+        """Can we start container and check sanity test?"""
+        base_url = "http://cl-binary-transformers-and-extractors:80"
```
Comment (mlissner) [inline cl/scrapers/tests.py:454]: Pull from settings.

File: cl/scrapers/tests.py [hunk 3]
```diff
@@ -438,3 +443,41 @@ def test_process_audio_file(self):
...
+        response = requests.get(url=base_url, timeout=60).json()
```
Comment (mlissner) [inline cl/scrapers/tests.py:455]: Tests should be fast, especially the heartbeat. We don't want to wait a minute if it's down. Give it a second, maybe two.

File: cl/scrapers/tests.py [hunk 4]
```diff
@@ -438,3 +443,41 @@ def test_process_audio_file(self):
...
+            "reagan",
+            "intelligence",
+            "october",
+            "tax court",
+            "fidelity",
+            "indiana",
+            "vermont",
+        ]
+        opinions = sorted(iglob(os.path.join(self.path, "*")))
+        for filepath, test_string in zip(opinions, test_strings):
```
Comment (mlissner) [inline cl/scrapers/tests.py:474]: Rework your data structure here to pair the words with the correct files. The whole thing works b/c of the order of the list and the names of the files, but that's way too fragile. Gotta be a better way.

File: cl/search/management/commands/cl_count_recap_pages.py [hunk 5]
```diff
@@ -1,8 +1,7 @@
 from collections import Counter
```
Comment (mlissner) [inline cl/search/management/commands/cl_count_recap_pages.py:1]: This whole file can be deleted. If we need it again, we can make it again. It's cruft.

File: docker/django/Dockerfile [hunk 6]
```diff
@@ -4,27 +4,11 @@ WORKDIR /usr/src/app
...
-        --assume-yes \
-        --quiet=2 \
-        `# Document extraction and OCR tools` \
-        antiword docx2txt ghostscript libwpd-tools poppler-utils \
-        `# Audio extraction/manipulation tools` \
-         ffmpeg \
-        `# Image & OCR tools` \
-        imagemagick tesseract-ocr tesseract-ocr-eng \
-        `# Other dependencies` \
-        libffi-dev libxml2-dev libxslt-dev python-dev
```
Comment (mlissner) [inline docker/django/Dockerfile:21]: I mean, this is beautiful to see.

File: docker/task-server/Dockerfile [hunk 7]
```diff
@@ -1,45 +1,15 @@
...
 ## Needs to be two commands so second one can use variables from first.
 ENV \
     CELERY_TASKS_DIR=/opt/celery \
     CELERY_USER_ID=33
 ENV \
     PYTHONPATH="${PYTHONPATH}:${CELERY_TASKS_DIR}"
 
-RUN mkdir /var/log/courtlistener && \
-    chmod 777 /var/log/courtlistener
-
```
Comment (mlissner) [inline docker/task-server/Dockerfile:42]: You're sure we can ditch this?

File: docker/courtlistener/.env [hunk 8]
```diff
@@ -0,0 +1,2 @@
+JUDGE_PICS_PATH=/usr/local/lib/python2.7/site-packages/judge_pics
+SEAL_ROOKERY_PATH=/usr/local/lib/python2.7/site-packages/seal_rookery
```
Comment (mlissner) [inline docker/courtlistener/.env:2]: Still no luck getting these installed in /opt?

**Review by mlissner — CHANGES_REQUESTED** [review]
OK, most everything makes sense here and only a couple larger comments. 

Should this include dropping dependencies from requirements.txt? Did you take a look in there?

File: docker/django/Dockerfile (hunk 6 above)
Comment (flooie) [reply docker/django/Dockerfile:21]: Indeed

File: docker/task-server/Dockerfile (hunk 7 above)
Comment (flooie) [reply docker/task-server/Dockerfile:42]: If my understanding of docker is right- its inherited line for line from the Django image

File: docker/courtlistener/.env (hunk 8 above)
Comment (flooie) [reply docker/courtlistener/.env:2]: correct.

File: cl/people_db/tasks.py [hunk 9]
```diff
@@ -22,37 +24,18 @@ def make_png_thumbnail_for_instance(
...
-    p = subprocess.Popen(
-        command, close_fds=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-    )
-    stdout, stderr = p.communicate()
-    if p.returncode != 0:
+    filepath = getattr(item, file_attr).path
+
+    response = generate_thumbnail(filepath)
+    err = response.headers["err"]
+    if err:
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/people_db/tasks.py:31]: this is a mistake on my part.

File: docker/courtlistener/.env (hunk 8 above)
Comment (mlissner) [reply docker/courtlistener/.env:2]: What's going wrong?

File: docker/task-server/Dockerfile (hunk 7 above)
Comment (mlissner) [reply docker/task-server/Dockerfile:42]: Yes, OK, that does make sense. Awesome.

File: cl/scrapers/docker_helpers.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def get_page_count(path, ext=None):
+    """Get page count from document
+
+    Sends file to binary-transformers-and-extractors and returns a json
+    object containing the pg count
+
+    :param path: path of the document
+    :param ext: File extension
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:84]: I was hoping to add doc and docx so I left it in for now.  The extension though could be determined easily from BTE

File: cl/scrapers/docker_helpers.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
+import os
+
+import requests
+from requests import Timeout
+
+base_url = "http://cl-binary-transformers-and-extractors:80"
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:6]: √

File: cl/scrapers/docker_helpers.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    service = "%s/%s" % (base_url, "extract_doc_content")
+    do_ocr = do_ocr
+    with open(path, "rb") as file:
+        f = file.read()
+    try:
+        return requests.post(
+            url=service,
+            files={"file": (os.path.basename(path), f)},
+            params={"do_ocr": do_ocr},
+            timeout=3600,
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:25]: √

File: cl/scrapers/docker_helpers.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    :return:
+    """
+    service = "%s/%s" % (base_url, "make_png_thumbnail")
+    with open(path, "rb") as file:
+        f = file.read()
+    try:
+        response = requests.post(
+            url=service,
+            files={"file": (os.path.basename(path), f)},
+            timeout=300,
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:68]: √

File: cl/scrapers/tasks.py [hunk 14]
```diff
@@ -295,27 +97,11 @@ def extract_doc_content(pk, do_ocr=False, citation_jitter=False):
...
-            "on opinion: %s****" % (extension, opinion)
-        )
-        return
-
+    response = document_extract(path, do_ocr)
+    content = response["content"]
+    success = response["err"]
     # Do page count, if possible
-    opinion.page_count = get_page_count(path, extension)
+    opinion.page_count = get_page_count(path)["pg_count"]
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/tasks.py:104]: yeah, I made it an optional variable to leave it out of the new calls and drift off into the ether for old calls until we added page count for other doc types

File: cl/scrapers/tasks.py [hunk 15]
```diff
@@ -374,24 +159,19 @@ def extract_recap_pdf(pks, skip_ocr=False, check_if_needed=True):
...
+        content = response["content"]
+        err = response["err"]
+
+        if err == Timeout:
+            content = u"Unable to extract document content."
+            rd.ocr_status = RECAPDocument.OCR_FAILED
+        elif len(content.strip()) == 0:
+            rd.ocr_status = RECAPDocument.OCR_NEEDED
         else:
-            rd.ocr_status = RECAPDocument.OCR_UNNECESSARY
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/tasks.py:394]: √

File: cl/scrapers/tasks.py (hunk 1 above)
Comment (flooie) [reply cl/scrapers/tasks.py:282]: I agree it should go in there - it was jus so intertwined with saving code that I didn't want to do it until I had this first draft reviewed

File: cl/scrapers/tests.py (hunk 2 above)
Comment (flooie) [reply cl/scrapers/tests.py:454]: √

File: cl/scrapers/tests.py (hunk 3 above)
Comment (flooie) [reply cl/scrapers/tests.py:455]: √

File: cl/scrapers/tests.py (hunk 4 above)
Comment (flooie) [reply cl/scrapers/tests.py:474]: I would prefer to just add the answers to a test assets directory and compare the results.  Perhaps you could take a look at the tests in the binary transformers extractors image repo and see if you like those tests?

File: .github/workflows/black.yml [hunk 16]
```diff
@@ -2,9 +2,11 @@ name: black
 
 on:
   push:
-    branches: [master, ]
+    branches:
+      - '**'
```
[the remark this answers is not in the document]
Comment (flooie) [reply .github/workflows/black.yml:6]: it looked like this 

`on: [push, pull_request]`

before the most recent update in august.  I'm not sure why it stopped working though - and I thought this maintained the most flexibility and style with what you were going for in the most recent change.  This is essentially regex - in case we want to exclude or include branches in the future.

File: docker/courtlistener/.env (hunk 8 above)
Comment (flooie) [reply docker/courtlistener/.env:2]: I can install into opt but I still need the same path to link to python2.7 in Django.

File: cl/recap/tasks.py [hunk 17]
```diff
@@ -306,7 +307,9 @@ def process_recap_pdf(self, pk):
 
             # Do page count and extraction
             extension = rd.filepath_local.path.split(".")[-1]
-            rd.page_count = get_page_count(rd.filepath_local.path, extension)
+            rd.page_count = get_page_count(rd.filepath_local.path, extension)[
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/recap/tasks.py:310]: Agreed.

File: cl/scrapers/docker_helpers.py [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+base_url = "http://cl-binary-transformers-and-extractors:80"
+
+
+def document_extract(path, do_ocr=False):
+    """Extract document content.
+
+    :param path:
+    :param do_ocr:
+    :return:
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:13]: Can I say draft?

File: cl/scrapers/tasks.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/tasks.py:394]: I wasnt sure it was necessary, we could keep it but its not used anywhere else as far as I can tell

File: cl/scrapers/docker_helpers.py [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+    try:
+        return requests.post(
+            url=service,
+            files={"file": (os.path.basename(path), f)},
+            timeout=300,
+        ).json()
+    except Timeout:
+        return {"err": Timeout}
+    except:
+        return {"err": "Unknown error occurred"}
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:97]: I was worried this is how you'd feel.  Sounds good though.

File: cl/scrapers/docker_helpers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/scrapers/docker_helpers.py:82]: I don't see us pulling those page counts anywhere on our roadmap. If we ever get to doing them, we can add the extension parameter back in.

File: cl/scrapers/tests.py (hunk 4 above)
Comment (mlissner) [reply cl/scrapers/tests.py:473]: Mind pointing me at a specific one in the transformers so I'm sure what you're suggesting? (Sorry if my earlier comment was a bit terse.)

File: cl/scrapers/tasks.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/scrapers/tasks.py:104]: Oh, I missed that it was optional. That clarifies.

File: cl/scrapers/tasks.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/scrapers/tasks.py:394]: You're right, I don't think it's used by our templates or anything, but I think it's still a good status to know when thinking about OCR stuff. It's possible our API and replication folks use it too, but not sure there.

File: cl/scrapers/tasks.py (hunk 1 above)
Comment (mlissner) [reply cl/scrapers/tasks.py:282]: Excellent.

File: cl/scrapers/docker_helpers.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/scrapers/docker_helpers.py:13]: Ha, sorry. I meant the comment mostly as a bookmark so we remember to catch it during later rounds of review.

File: cl/scrapers/docker_helpers.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
+import os
```
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:1]: How about something like microservice_calls or bte_calls

File: cl/scrapers/docker_helpers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:82]: √

[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:82]: removed

File: cl/scrapers/tests.py (hunk 4 above)
Comment (flooie) [reply cl/scrapers/tests.py:473]: @mlissner https://github.com/freelawproject/courtlistener-binary-transformers-and-extractors/blob/b53013c42be792c28ce30b81e02158f7ee7c20cc/tests/tests.py#L113. this entire class

File: cl/scrapers/docker_helpers.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:90]: Something like this @mlissner 

```
def generate_thumbnail(path):
    """Generate a thumbnail of page 1 of a PDF

    :param path:
    :return:
    """
    service = "%s/%s" % (settings.BTE_URL, "make_png_thumbnail")
    with open(path, "rb") as file:
        f = file.read()
    return requests.post(
        url=service,
        files={"file": (os.path.basename(path), f)},
        timeout=60,
    ).content

```

& 

```
    try:
        thumbnail = generate_thumbnail(filepath)
    except Timeout:
        logger.error("Thumnail generation failed via timeout.")
        item.thumbnail_status = THUMBNAIL_STATUSES.FAILED
        item.save()
        return item.pk
    except Exception as e:
        logger.error(
            "Catch all exception occurred during thumbnail generation.  See %s"
            % str(e)
        )
        item.thumbnail_status = THUMBNAIL_STATUSES.FAILED
        item.save()
        return item.pk
```

File: cl/people_db/tasks.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/people_db/tasks.py]: √

[the remark this answers is not in the document]
Comment (flooie) [reply cl/people_db/tasks.py]: I think this should be resolved in the above code.

File: cl/scrapers/docker_helpers.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py:90]: @mlissner just re-upping this.

[the remark this answers is not in the document]
Comment (mlissner) [reply cl/scrapers/docker_helpers.py:90]: I think you can avoid duplicating the item.save() and return statements, but yeah, that looks like the right pattern to me.

File: cl/scrapers/docker_helpers.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (mlissner) [reply cl/scrapers/docker_helpers.py:1]: How about transformer_extractor_utils in an effort to be more explicit?

File: .github/workflows/black.yml (hunk 16 above)
[the remark this answers is not in the document]
Comment (mlissner) [reply .github/workflows/black.yml:6]: LGTM.

File: cl/scrapers/tests.py (hunk 4 above)
Comment (mlissner) [reply cl/scrapers/tests.py:473]: Yeah, I think the bte approach is a step in the right direction (it doesn't depend on the alphabetical ordering of a directory to work), but I still find it pretty tricky to work with. You have to go from `DocumentConversionTests` to its super class, `DockerTestBase`, to a json file it creates, through the code that converts it, then to the key in that file, and finally you know what the test is being compared to. Wouldn't it be easier if those tests just had something like juriscraper does:

    some_file.pdf
    some_file_result.txt

And then you just grab the contents of `some_file_result.{txt,json,etc}` when doing your assertion? 

---

For these tests though, it occurs to me (belatedly, sorry) that the only thing we need to test is that the boundaries work. That's the beauty of the microservices. In BTE, we test that conversions happen properly; we don't need to *also* test that here every time we make a one-line tweak to the code. 

All we need to make sure is that they're properly compatible at the boundaries. Maybe just send one really basic request to each endpoint, and make sure it returns a 200 or something? (I'm assuming the endpoints will fail if they get parameters they don't like?)

File: docker/courtlistener/.env (hunk 8 above)
Comment (mlissner) [reply docker/courtlistener/.env:2]: I see. Thanks for trying.

File: cl/scrapers/docker_helpers.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py]: √

File: cl/scrapers/docker_helpers.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (flooie) [reply cl/scrapers/docker_helpers.py]: I think this should be resolved.

File: cl/scrapers/tasks.py (hunk 1 above)
Comment (flooie) [reply cl/scrapers/tasks.py]: This has been moved out of CL - and allowed for some cleanup.

File: cl/cleanup/management/commands/cleanup_636_pacer_leading_zeros.py [hunk 21]
```diff
@@ -11,7 +11,8 @@
     get_local_document_url_from_path,
 )
 from cl.search.models import RECAPDocument, Docket
-from cl.scrapers.tasks import get_page_count, extract_recap_pdf
+from cl.scrapers.tasks import extract_recap_pdf
```
Comment (mlissner) [inline cl/cleanup/management/commands/cleanup_636_pacer_leading_zeros.py:14]: Let's just remove this whole file. I tire of our cruft.

Comment (flooie) [reply cl/cleanup/management/commands/cleanup_636_pacer_leading_zeros.py]: √ removed

File: cl/people_db/tasks.py [hunk 22]
```diff
@@ -22,37 +25,28 @@ def make_png_thumbnail_for_instance(
...
+
+    try:
+        thumbnail_resp = generate_thumbnail(filepath)
+    except Timeout:
+        logger.error("Thumbnail generation failed via timeout.")
+    except Exception as e:
+        logger.error(
+            "Catch all exception occurred during thumbnail generation.  See %s"
+            % str(e)
+        )
```
Comment (mlissner) [inline cl/people_db/tasks.py:38]: You might as well just not have this catch all. WIthout it, an error goes to Sentry with full stacktrace and other great stuff. With it, we only get the log line.

```suggestion
```

File: cl/people_db/tasks.py [hunk 23]
```diff
@@ -22,37 +25,28 @@ def make_png_thumbnail_for_instance(
...
+        if "thumbnail_resp" not in locals():
+            item.thumbnail_status = THUMBNAIL_STATUSES.FAILED
+            item.save()
+            return item.pk
 
     item.thumbnail_status = THUMBNAIL_STATUSES.COMPLETE
     filename = "%s.thumb.%s.jpeg" % (pk, max_dimension)
-    item.thumbnail.save(filename, ContentFile(stdout))
+    item.thumbnail.save(
+        filename, ContentFile(literal_eval(thumbnail_resp["content"]))
```
Comment (mlissner) [inline cl/people_db/tasks.py:43]: Please clean up.

File: cl/scrapers/transformer_extractor_utils.py [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+from django.core import serializers
+
+
+def document_extract(path=None, file_content=None, do_ocr=False):
+    """Extract document content.
+
+    :param path:
+    :param file_content:
+    :param do_ocr:
+    :return:
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:16]: Missed these, looks like.

File: cl/scrapers/transformer_extractor_utils.py [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+    if path is not None:
+        with open(path, "rb") as file:
+            file_content = file.read()
+    elif file_content is None:
+        return {"err": "File not supplied"}
+
+    if path is not None:
+        file_name = os.path.basename(path)
+    else:
+        file_name = os.path.basename(file_content.name)
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:28]: I'm...confused what kind of object file_content is. If it's just a str of data, it wouldn't have a `name` property, right?

File: cl/scrapers/transformer_extractor_utils.py [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+        with open(tmp.name, "rb") as file:
+            af = file.read()
+    return af
+
+
+def convert_and_clean_audio(af):
+    """Convert audio file to MP3 and add metadata
+
+    :param path:
+    :return:
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:64]: Another.

File: cl/scrapers/transformer_extractor_utils.py [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+
+    files = {
+        "file": ("the_audio.wav", wav),
+        "af": ("the_json.json", audio_json),
+    }
+    return requests.post(
+        url=service,
+        files=files,
+        timeout=60 * 60,
+    )
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:79]: Do a little research on how to properly post JSON data using requests. We shouldn't be sending audio metadata this way, the BTE should instead be configured to receive POST parameters.

File: cl/scrapers/transformer_extractor_utils.py [hunk 28]
```diff
@@ -0,0 +1,186 @@
...
+def convert_and_clean_audio(af):
+    """Convert audio file to MP3 and add metadata
+
+    :param path:
+    :return:
+    """
+    service = "%s/%s/%s" % (settings.BTE_URL, "convert", "audio")
+    audio_json = serialize_audio_object(af)
+    with open(af.local_path_original_file.path, "rb") as wma_file:
+        wav = wma_file.read()
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:69]: These won't always be wav files, I don't think.

File: cl/scrapers/transformer_extractor_utils.py [hunk 29]
```diff
@@ -0,0 +1,186 @@
...
+        0
+    ]["fields"]
+    af_dict["docket"] = docket_dict
+    af_dict["docket"]["court"] = court_dict
+    with NamedTemporaryFile(suffix=".json") as tmp:
+        with open(tmp.name, "w") as json_data:
+            json.dump(af_dict, json_data)
+        with open(tmp.name, "rb") as file:
+            af = file.read()
+    return af
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:57]: I think this will be obviated by doing the POST correctly, but there's a lot going on here. I guess I'll just note two things:

1. You can always do `object.__dict__()` to convert them to a dict. I think that's easier than the hoops you're going through here.

2. You're creating JSON data from a Python object be `json.dump`ing it to disk, and then reading it from disk. Check out `json.dumps`, where the `s` is short for string (I believe). It's the cousin of `loads`. 

Apologies if I'm missing something here.

File: cl/scrapers/transformer_extractor_utils.py [hunk 30]
```diff
@@ -0,0 +1,186 @@
...
+        url=service,
+        files=files,
+        timeout=60 * 60,
+    )
+
+
+def generate_thumbnail(path=None, file_content=None):
+    """Generate a thumbnail of page 1 of a PDF
+
+    :param path:
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:85]: Docstring, please.

File: cl/scrapers/transformer_extractor_utils.py [hunk 31]
```diff
@@ -0,0 +1,186 @@
...
+    :return:
+    """
+    service = "%s/%s" % (settings.BTE_URL, "make_png_thumbnail")
+    if path is not None:
+        with open(path, "rb") as file:
+            file_content = file.read()
+    elif file_content is None:
+        return {"err": "File not supplied"}
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:93]: It feels worth it to just decide if this function takes data (most flexible) or a path to data (less so), and not support both.

Comment (mlissner) [reply cl/scrapers/transformer_extractor_utils.py:93]: Probably applies to other functions around here too?

Comment (flooie) [reply cl/scrapers/transformer_extractor_utils.py:93]: sounds good - I as you see wanted to split the baby

File: cl/scrapers/transformer_extractor_utils.py [hunk 32]
```diff
@@ -0,0 +1,186 @@
...
+
+
+def get_page_count(path=None, file_content=None):
+    """Get page count from document
+
+    Sends file to binary-transformers-and-extractors and returns a json
+    object containing the pg count
+
+    :param path: path of the document
+    :param pdf_bytes:
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:109]: Clean up, please.

File: cl/scrapers/transformer_extractor_utils.py [hunk 33]
```diff
@@ -0,0 +1,186 @@
...
+
+    service = "%s/%s/%s" % (
+        settings.BTE_URL,
+        "financial_disclosure",
+        endpoint,
+    )
+    if file_content is not None:
+        files = {"file": ("some.pdf", file_content)}
+    else:
+        files = None
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:146]: When would you want `files == None`?

File: cl/scrapers/transformer_extractor_utils.py [hunk 34]
```diff
@@ -0,0 +1,186 @@
...
+        return {"err": "File not supplied"}
+
+    return requests.post(
+        url=service,
+        files={"file": ("some.pdf", file_content)},
+        timeout=60,
+    ).json()
+
+
+def extract_jw_financial_document(
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:126]: The function name implies it does `jw` disclosures, but by default it does not. Maybe just tweak the name?

File: cl/scrapers/transformer_extractor_utils.py [hunk 35]
```diff
@@ -0,0 +1,186 @@
...
+    url=None, file_content=None, judicial_watch=False
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:127]: `url` isn't used
`file_content` shouldn't have a default param if it's a required param (I feel like this probably applies to other areas in here too — worth a quick look?)

File: cl/scrapers/transformer_extractor_utils.py [hunk 36]
```diff
@@ -0,0 +1,186 @@
...
+        files = None
+    response = requests.post(
+        url=service,
+        files=files,
+        params={"url": url},
+    )
+    return response
+
+
+def make_pdf_from_single_image(aws_path=None, file_content=None):
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:155]: `file_path` isn't used. This is one of the reasons not to give default params to method signatures, because it allows cruft to stick around.

File: cl/scrapers/transformer_extractor_utils.py [hunk 37]
```diff
@@ -0,0 +1,186 @@
...
+        "single_image",
+    )
+    response = requests.post(
+        url=service,
+        params={"aws_path": aws_path},
+    )
+    return response
+
+
+def make_pdf_from_multiple_images(aws_path=None, file_content=None):
```
Comment (mlissner) [inline cl/scrapers/transformer_extractor_utils.py:172]: `file_content` unused.

File: docker/courtlistener/docker-compose.yml [hunk 38]
```diff
@@ -45,31 +48,37 @@ services:
     # Task Server
     cl-celery:
         container_name: cl-celery
-        image: freelawproject/task-server:latest
+        image: freelawproject/task-server:bte_beta
```
Comment (mlissner) [inline docker/courtlistener/docker-compose.yml:51]: We want this back on `latest`, right?

File: docker/courtlistener/docker-compose.yml [hunk 39]
```diff
@@ -45,31 +48,37 @@ services:
...
             - ${CL_POSTGRES_RUN_DIR:-/var/run/postgresql}:/var/run/postgresql
             - ${CL_BASE_DIR:-../..}/cl/assets/media:/opt/celery/cl/assets/media
+            - cl-judge-pics:${JUDGE_PICS_PATH}
         networks:
             - cl_net_overlay
 
     cl-django:
         container_name: cl-django
-        image: freelawproject/courtlistener-django:latest
+        image: freelawproject/courtlistener-django:bte_beta
```
Comment (mlissner) [inline docker/courtlistener/docker-compose.yml:67]: Ditto here.

File: cl/settings/10-public.py [hunk 40]
```diff
@@ -579,3 +579,9 @@ def skip_unreadable_post(record):
     AWS_S3_CUSTOM_DOMAIN = "%s.s3.amazonaws.com" % AWS_STORAGE_BUCKET_NAME
 
 CLOUDFRONT_DOMAIN = ""
+
+####################################
+# Binary Transformers & Extractors #
+####################################
+
+BTE_URL = "http://cl-binary-transformers-and-extractors:80"
```
Comment (mlissner) [inline cl/settings/10-public.py:587]: I'm a little surprised you can run this on port 80, but I guess it works since it's inside the docker network. But I think we should pick and use a high-numbered port for this since the low ones are essentially spoken for. I'm no expert here, but can you think about it and look at port allocations and see which might make sense? Ideally, we just don't want it to be an existing port or one we may wind up at someday, and if it somehow makes sense, that's a benefit. Like 2020 might not be crazy, I dunno.

File: docker/task-server/Dockerfile [hunk 41]
```diff
@@ -1,45 +1,15 @@
-FROM python:2.7
+#Inherit our django image which has the same pip requirements start the celery container.
+FROM freelawproject/courtlistener-django:bte_beta
```
Comment (mlissner) [inline docker/task-server/Dockerfile:2]: Again here.

File: cl/scrapers/tasks.py [hunk 42]
```diff
@@ -374,24 +154,19 @@ def extract_recap_pdf(pks, skip_ocr=False, check_if_needed=True):
...
-                    content = u"Unable to extract document content."
-                    rd.ocr_status = RECAPDocument.OCR_FAILED
-            else:
-                content = u""
-                rd.ocr_status = RECAPDocument.OCR_NEEDED
+        response = document_extract(path, do_ocr=True)
+        content = response["content"]
+        err = response["err"]
+
+        if err == Timeout:
```
Comment (mlissner) [inline cl/scrapers/tasks.py:163]: This doesn't look right.

File: cl/scrapers/tasks.py [hunk 43]
```diff
@@ -374,24 +154,19 @@ def extract_recap_pdf(pks, skip_ocr=False, check_if_needed=True):
             # hasn't disabled early abortion.
             processed.append(pk)
             continue
+
```
Comment (mlissner) [inline cl/scrapers/tasks.py:157]: It looks like you tweaked the logic here so that OCR_UNNECESSARY is never set, and so that skip_ocr is never checked. Can you please review?

File: cl/scrapers/tasks.py [hunk 44]
```diff
@@ -401,218 +176,37 @@ def extract_recap_pdf(pks, skip_ocr=False, check_if_needed=True):
...
-                e.returncode,
-                e.output,
-                traceback.format_exc(),
-            )
+        response = convert_and_clean_audio(af).json()
+        cf = ContentFile(response["content"])
+        file_name = trunc(best_case_name(af).lower(), 72) + "_cl.mp3"
+        af.file_with_date = af.docket.date_argued
+        af.local_path_mp3.save(file_name, cf, save=False)
+        af.duration = response["duration"]
```
Comment (mlissner) [inline cl/scrapers/tasks.py:191]: Should these lines go after the except clause, before the af.processing_complete line?

File: cl/scrapers/tests.py [hunk 45]
```diff
@@ -99,9 +104,12 @@ def test_txt_extraction_with_bad_data(self):
             "search",
             "txt_file_with_no_encoding.txt",
         )
-        content, err = extract_from_txt(path)
+        response = document_extract(path)
+        content = response["content"]
+
         self.assertFalse(
-            err, "Error reported while extracting text from %s" % path
+            int(response["error_code"]),
```
Comment (mlissner) [inline cl/scrapers/tests.py:111]: What is error_code? Can we use the HTTP status instead?

File: cl/settings/10-public.py (hunk 40 above)
Comment (flooie) [reply cl/settings/10-public.py:587]: 2020, may never want to see that number again
